spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -178,47 +197,46 @@ else if (!returnedTypeIsNumber) {
 		structredQueryBuilder.setKind(this.datastorePersistentEntity.kindName());
 
 		boolean singularResult = !isCountingQuery && collectionType == null;
-		Iterable rawResults = getDatastoreTemplate()
+		DatastoreResultsIterable rawResults = getDatastoreTemplate()
 				.queryKeysOrEntities(
-						applyQueryBody(parameters, structredQueryBuilder, total, singularResult),
+						applyQueryBody(parameters, structredQueryBuilder, total, singularResult, null),
 						this.entityType);
 
 		Object result = StreamSupport.stream(rawResults.spliterator(), false).map(mapper).collect(collector);
 
 		if (this.tree.isDelete()) {
 			deleteFoundEntities(returnedTypeIsNumber, rawResults);
 		}
-
-		return (this.tree.isExistsProjection() || isCountingQuery) ? result
-				: convertResultCollection(result, collectionType);
+		boolean countingOrExistsQuery = this.tree.isExistsProjection() || isCountingQuery;
+		if (!countingOrExistsQuery) {
+			this.cursor = rawResults.getCursor();
+			return convertResultCollection(result, collectionType);
+		}
+		else {
+			return result;
+		}
 	}
 
 	private Slice executeSliceQuery(Object[] parameters) {
 		EntityQuery.Builder builder = StructuredQuery.newEntityQueryBuilder()
 				.setKind(this.datastorePersistentEntity.kindName());
-		StructuredQuery query = applyQueryBody(parameters, builder, false, false);
-		List items = this.datastoreTemplate.query((query), (x) -> x);
-		Integer limit = (query.getLimit() == null) ? null : query.getLimit() - 1;
-
-		boolean exceedsLimit = false;
-		if (limit != null) {
-			//for slice queries we retrieve one additional item to check if the next slice exists
-			//the additional item will not be converted on read
-			exceedsLimit = items.size() > limit;
-			if (exceedsLimit) {
-				items = items.subList(0, limit);
-			}
-		}
+		StructuredQuery query = applyQueryBody(parameters, builder, false, false, null);
+		DatastoreResultsIterable<?> resultList = this.datastoreTemplate.queryKeysOrEntities(query, this.entityType);
 
 		ParameterAccessor paramAccessor = new ParametersParameterAccessor(getQueryMethod().getParameters(), parameters);
-		Pageable pageable = paramAccessor.getPageable();
-		List entities = (List) this.datastoreTemplate
-				.convertEntitiesForRead(items.iterator(), this.entityType).stream()
-				.map((o) -> this.processRawObjectForProjection((T) o)).collect(Collectors.toList());
-		return new SliceImpl(entities, pageable, exceedsLimit);
+
+		Pageable pageable = CursorPageable.from(paramAccessor.getPageable(), resultList.getCursor(), null);
+
+		EntityQuery.Builder builderNext = StructuredQuery.newEntityQueryBuilder()
+				.setKind(this.datastorePersistentEntity.kindName());
+		StructuredQuery queryNext = applyQueryBody(parameters, builderNext, false, true, resultList.getCursor());
+		List datastoreResultsList = this.datastoreTemplate.query(queryNext, x -> x);
```

Comment:
Can be boolean instead of `List`.

Reply:
has to be a list, because `query` returns a list